src/service/crypto-verifier-impl.js

```diff
@@ -0,0 +1,198 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../log';
+import {getService} from '../service';
+import {
+  encodeByteArray,
+} from
+  '../../third_party/closure-library/encode-byte-array-trampoline-generated';
+
+
+const TAG_ = 'CryptoVerifier';
+
+const isWebkit = window.crypto && 'webkitSubtle' in window.crypto;
+const crossCrypto = isWebkit ? window.crypto.webkitSubtle :
+                               window.crypto.subtle;
+
+const VERSION = 0x00;
+
+/**
+ * A service that provides cryptographic signature verification for use
+ * within AMP.
+ */
+class CryptoVerifier {
+  /**
+   * Verifies RSA signature corresponds to the data given a list of public keys.
+   * @param {!Uint8Array} the data that was signed.
+   * @param {!Uint8Array} the RSA signature.
+   * @param {{e: !Uint8Array, n:!Uint8Array, keyHash: !Uint8Array }} the RSA
+   *     public keys. If keyHash in each public key is null, this function will
+   *     compute it.
+   * @return {!Promise<!boolean>} whether the signature is valid for one of
+   *     the public keys.
+   */
+  verify(data, signature, rsaPubKeys) {
```

Comment:
Looking at this purely from the API, we have two different return types:

- A Promise
  - This promise may resolve to true or false
  - This promise may reject with due to some crypto error
- Throws an error because `crossCrypto` is undefined.

Generally, mixing a Promise return value and throwing an error [is bad](http://www.2ality.com/2016/03/promise-rejections-vs-exceptions.html). Would it be better here to return a rejected promise with some "no suitable crypto library" error?